pkg/skippeers/skippeers.go

```diff
@@ -5,41 +5,119 @@
 package skippeers
 
 import (
+	"container/heap"
+	"math"
 	"sync"
 	"time"
 
 	"github.com/ethersphere/bee/pkg/swarm"
 )
 
+const maxDuration int64 = math.MaxInt64
+
 type List struct {
 	mtx sync.Mutex
 
+	minHeap *timeHeap
+
+	durC chan time.Duration
+	quit chan struct{}
 	// key is chunk address, value is map of peer address to expiration
 	skip map[string]map[string]int64
+
+	wg sync.WaitGroup
 }
 
 func NewList() *List {
-	return &List{
-		skip: make(map[string]map[string]int64),
+	l := &List{
+		minHeap: &timeHeap{},
+		skip:    make(map[string]map[string]int64),
+		durC:    make(chan time.Duration),
+		quit:    make(chan struct{}),
 	}
+
+	l.wg.Add(1)
+	go l.worker()
+
+	return l
+}
+
+func (l *List) worker() {
+
+	defer l.wg.Done()
+
+	var (
+		timer  *time.Timer
+		timerC <-chan time.Time
+	)
+
+	for {
+		select {
+		case dur := <-l.durC:
+			if timer == nil {
+				timer = time.NewTimer(dur)
+			} else {
+				if !timer.Stop() {
+					<-timer.C
+				}
+				timer.Reset(dur)
+			}
+			timerC = timer.C
+		case <-timerC:
+			l.PruneExpiresAfter(0)
+			l.mtx.Lock()
+			if l.minHeap.Length() > 0 {
+				timer.Reset(time.Until(time.Unix(0, l.minHeap.First())))
+			}
+			l.mtx.Unlock()
+		case <-l.quit:
+			return
+		}
+	}
+
 }
 
 func (l *List) Add(chunk, peer swarm.Address, expire time.Duration) {
 	l.mtx.Lock()
 	defer l.mtx.Unlock()
 
+	t := time.Now().Add(expire).UnixNano()
+
+	heap.Push(l.minHeap, t)
+	min := l.minHeap.First()
+
+	if t == min { // pushed the most recent expiration
+		select {
+		case l.durC <- time.Until(time.Unix(0, min)):
+		case <-l.quit:
+		}
+	}
+
+	if _, ok := l.skip[chunk.ByteString()]; !ok {
+		l.skip[chunk.ByteString()] = make(map[string]int64)
+	}
+
+	l.skip[chunk.ByteString()][peer.ByteString()] = t
+}
+
+func (l *List) AddForever(chunk, peer swarm.Address) {
+	l.mtx.Lock()
+	defer l.mtx.Unlock()
+
+	heap.Push(l.minHeap, maxDuration)
```

Comment:
Do we need to add this to the heap? I dont think these entries will be used to trigger the pruning, so it would be better if we just dont add them to the heap. This would even reduce some unnecessary ops while adding to heap etc.